src/ra_utility.h

```diff
@@ -172,6 +176,77 @@ _NODISCARD inline auto __cdecl tstrlen(_In_z_ LPCTSTR _Str) noexcept
+/// <typeparam name="CharT">
+///   A character type, specify this explicitly to use a different character set than configured.
+///   Valid character types are <c>char</c>, <c>signed char</c>, and <c>wchar_T</c>. Using any other
+///   type will disable this function.
+/// </typeparam>
+/// <param name="buffer">
+///   Pointer to a character string to write to, if this is <c>nullptr</c>, the operation is
+///   undefined.
+/// </param>
+/// <param name="bufsz">
+///   Up to <c><paramref name="bufsz" /> - 1</c> characters may be written, plus the null terminator.
+///   If the size is <c>0</c> or greater than <c>RSIZE_MAX</c>, the operation is undefined.
+/// </param>
+/// <param name="format">
+///   Pointer to a null-terminated multibyte string or wide-character string specifying how to
+///   interpret the data. The format string consists of ordinary multibyte characters(except %),
+///   which are copied unchanged into the output stream, and conversion specifications. The %n
+///   specifier is considered undefined.
+/// </param>
+/// <param name="...args">
+///   Forwarded arguments specifying data to print. If the number of arguments is more than the
+///   number of format specifiers present, the operation is undefined.
+/// </param>
+/// <returns>
+///   The number of characters written to <paramref name="buffer" /> excluding the null character.
+/// </returns>
+template<typename CharT = TCHAR, typename ...Args, typename = std::enable_if_t<is_char_v<CharT>>>
+_Success_(return >= 0) /*_NODISCARD*/ inline auto __cdecl
+stprintf_s(_Inout_updates_bytes_(bufsz)  CharT*       const __restrict buffer,
+           _In_range_(1, RSIZE_MAX)      rsize_t      const            bufsz,
+           _In_z_ _Printf_format_string_ const CharT* const __restrict format,
+           _In_                          Args&&...                     args)
+{
+    assert((buffer != nullptr) && (format != nullptr));
+    assert((bufsz > 0) && (bufsz <= RSIZE_MAX));
+    
+    if constexpr(std::is_same_v<CharT, char>)
+    {
+        [[maybe_unused]] std::string_view test{ format };
+        assert(test.find("%n") == std::string::npos);
+        [[maybe_unused]] const auto num_specs{ std::count(test.begin(), test.end(), '%') };
+        assert(num_specs == sizeof...(args));
+    }
+    else if constexpr(std::is_same_v<CharT, wchar_t>)
+    {
+        [[maybe_unused]] std::wstring_view test{ format };
+        assert(test.find(L"%n") == std::wstring::npos);
+        [[maybe_unused]] const auto num_specs{ std::count(test.begin(), test.end(), L'%') };
+        assert(num_specs == sizeof...(args));
+    }
+    
+    // MSVC doesn't seem to have constraint handlers like C11
+    if constexpr (std::is_same_v<CharT, char>)
+    {
+        const auto ret{ std::snprintf(buffer, bufsz, format, std::forward<Args>(args)...) };
+        assert(std::char_traits<char>::length(buffer) < bufsz);
+        assert(ret > 0);
```

Comment:
`std::char_traits<char_type>::length` is a constant expression, it shouldn't give any overhead.